Firebase/Messaging/FIRMessaging.m

```diff
@@ -141,40 +148,46 @@ @interface FIRMessaging ()<FIRMessagingClientDelegate, FIRMessagingReceiverDeleg
 /// which can happen if the user inadvertently calls `appDidReceiveMessage` along with us
 /// calling it implicitly during swizzling.
 @property(nonatomic, readwrite, strong) NSMutableSet *loggedMessageIDs;
+@property(nonatomic, readwrite, strong) id<FIRAnalyticsInterop> _Nullable analytics;
 
-- (instancetype)initWithInstanceID:(FIRInstanceID *)instanceID
-                      userDefaults:(NSUserDefaults *)defaults NS_DESIGNATED_INITIALIZER;
+@end
+
+// Messaging doesn't provide any functionality to other components,
+// so it provides a private, empty protocol that it conforms to and use it for registration.
+
+@protocol FIRMessagingInstanceProvider
+@end
 
+@interface FIRMessaging () <FIRMessagingInstanceProvider,
+                            FIRCoreConfigurable,
+                            FIRComponentRegistrant>
 @end
 
 @implementation FIRMessaging
 
 + (FIRMessaging *)messaging {
-  static FIRMessaging *messaging;
+  FIRApp *defaultApp = [FIRApp defaultApp];  // Missing configure will be logged here.
+  id<FIRMessagingInstanceProvider> messaging =
+      FIR_COMPONENT(FIRMessagingInstanceProvider, defaultApp.container);
+
   static dispatch_once_t onceToken;
   dispatch_once(&onceToken, ^{
-    messaging = [[FIRMessaging alloc] initPrivately];
-    [messaging start];
+    [(FIRMessaging *)messaging start];
   });
-  return messaging;
+  return (FIRMessaging *)messaging;
 }
 
-- (instancetype)initWithInstanceID:(FIRInstanceID *)instanceID
-                      userDefaults:(NSUserDefaults *)defaults {
+- (instancetype)initWithAnalytics:(nullable id<FIRAnalyticsInterop>)analytics {
```

Comment:
ah nice refactor. Thank you for doing it!

Reply:
It may be nice to keep the user defaults injection and instanceID injection here for testing purposes instead of relying on the two singletons in the internal method - we can inject the singletons in the `+componentsToRegister` call. 

Also, filed #1919 to move Messaging to `GULUserDefaults`.